src/bootstrap/doc.rs

```diff
@@ -595,26 +621,52 @@ impl Step for Rustc {
         // Only include compiler crates, no dependencies of those, such as `libc`.
         cargo.arg("--no-deps");
 
-        // Find dependencies for top level crates.
         let mut compiler_crates = HashSet::new();
-        for root_crate in &["rustc_driver", "rustc_codegen_llvm", "rustc_codegen_ssa"] {
-            compiler_crates.extend(
-                builder
-                    .in_tree_crates(root_crate, Some(target))
-                    .into_iter()
-                    .map(|krate| krate.name),
-            );
+
+        if paths.is_empty() {
+            // Find dependencies for top level crates.
+            for root_crate in &["rustc_driver", "rustc_codegen_llvm", "rustc_codegen_ssa"] {
+                compiler_crates.extend(
+                    builder
+                        .in_tree_crates(root_crate, Some(target))
+                        .into_iter()
+                        .map(|krate| krate.name),
+                );
+            }
+        } else {
+            for root_crate in paths {
+                if !builder.src.join("compiler").join(&root_crate).exists() {
+                    builder.info(&format!(
+                        "\tskipping - compiler/{} (unknown compiler crate)",
+                        root_crate
+                    ));
+                } else {
+                    compiler_crates.extend(
+                        builder
+                            .in_tree_crates(root_crate, Some(target))
+                            .into_iter()
+                            .map(|krate| krate.name),
+                    );
+                }
+            }
         }
 
+        let mut to_open = None;
         for krate in &compiler_crates {
             // Create all crate output directories first to make sure rustdoc uses
             // relative links.
             // FIXME: Cargo should probably do this itself.
             t!(fs::create_dir_all(out_dir.join(krate)));
             cargo.arg("-p").arg(krate);
+            to_open = Some(krate);
         }
 
         builder.run(&mut cargo.into());
+        // Let's open the first crate documentation page:
```

Comment:
This will open the last crate, not the first, right? Since to_open gets assigned in every iteration above?

Reply:
Oups indeed! Forgot to check if the variable was already set... Good catch!

Reply:
Actually I was also wrong above -- this is going to open a random crate pretty much, because the loop above is iterating a hashset. So just checking that it's none isn't quite so helpful.

I'm personally fine with just leaving it, I doubt people will care much -- my guess is almost no one uses --open more than once at the CLI (just refresh the browser instead).

Reply:
Well, let's keep as is then. In the end, the crate they want will be there in any case... :D